desc/equilibrium/equilibrium.py

```diff
@@ -836,60 +888,100 @@ def compute(
+        # Unless user asks, don't try to recompute stuff which are only dependencies
+        # of dep0d. Example, need R0. R0 <- A <- A(z) computable on dep0d grid.
+        # But A(z) in dep1dz and attempt to recompute on dep1dz grid will error
+        # without unique zeta idx defined, which mapped coordinate grids lack.
+        dep0d_deps = set(get_data_deps(dep0d, obj=p, has_axis=grid.axis.size))
+        # This filter is stronger than the name implies, but the false positives
+        # will still be computed correctly with the logic in compute.utils.compute.
+        just_dep0d_dep = lambda name: name in dep0d_deps and name not in names
+        dep1dr = {
             dep
             for dep in deps
-            if (data_index[p][dep]["coordinates"] == "r") and (dep not in data)
-        ]
-        dep1dz = [
+            if self.is_1dr(dep) and not just_dep0d_dep(dep) and dep not in need_src_deps
+        }
+        dep1dz = {
             dep
             for dep in deps
-            if (data_index[p][dep]["coordinates"] == "z")
-            and (dep not in data)
-            and dep not in ["phi", "zeta"]  # these don't need a special grid
-        ]
-
-        # whether we need to calculate 0d or 1d quantities on a special grid
-        calc0d = bool(len(dep0d))
-        calc1dr = bool(len(dep1dr))
-        calc1dz = bool(len(dep1dz))
-        if (  # see if the grid we're already using will work for desired qtys
-            (grid.L >= self.L_grid)
-            and (grid.M >= self.M_grid)
-            and (grid.N >= self.N_grid)
-        ):
+            if self.is_1dz(dep) and not just_dep0d_dep(dep) and dep not in need_src_deps
+            # These don't need a special grid, since the transforms are always
+            # built on the (rho, theta, zeta) coordinate grid.
+            and dep not in ["phi", "zeta"]
+        }
+
+        # Whether we need to calculate any dependencies on a special grid.
+        calc0d = bool(dep0d)
+        calc1dr = bool(dep1dr)
+        calc1dz = bool(dep1dz)
+        # If the grid samples the full volume, then it is sufficient.
+        if grid.L >= self.L_grid and grid.M >= self.M_grid and grid.N >= self.N_grid:
             if isinstance(grid, QuadratureGrid):
                 calc0d = calc1dr = calc1dz = False
             if isinstance(grid, LinearGrid):
                 calc1dr = calc1dz = False
+        else:
+            # Warn if best way to compute accurately is increasing resolution.
+            for dep in deps:
+                req = data_index[p][dep]["resolution_requirement"]
+                coords = data_index[p][dep]["coordinates"]
+                msg = lambda direction: colored(
+                    f"Dependency {dep} may require more {direction}"
+                    f" resolution to compute.",
+                    "yellow",
+                )
+                warnif(
+                    "r" in req and "r" in coords and grid.L < self.L_grid,
```

Comment:
why do we need to check both `req` and `coords`? Shouldn't `req` be enough?

Reply:
Suppose you want to compute `A(z)` which has coords="z". Right now `A(z)` has empty string as resolution requirement, since `coordinates"z"` implicitly implies that you need radial and poloidal resolution, but we could just as well explicitly  specify that `resolution_requirement="rt"`.

So:
```python
name="A(z)"
coordinates="z"              # output is function of toroidal coordinate only
resolution_requirement="rt"  # radial and poloidal resolution needed

footgun_grid = Grid(np.array([[0, 0, zeta]]))  # 1d grid in zeta
# common footgun error to pass in grid that has only the desired output evaluation points
data = eq.compute(name, grid=footgun_grid)
```
Since the point of the logic in `eq.compute()` is to save the user from such errors, we will override the grid and compute `A(z)` on a grid with radial resolution and (losslessly)  interpolate back to the original grid. When we get to the following chunk of code
```python
# don't need to warn because we won't use this grid to calculate anyway.
warnif(
    # True and False and True
    "r" in req and "r" in coords and footgun_grid.L < eq.L_grid,
    ResolutionWarning,
    msg("radial"),
)
warnif(
    # True and False and True
    "t" in req and "t" in coords and footgun_grid.M < eq.M_grid,
    ResolutionWarning,
    msg("poloidal"),
)
```
The additional check that `"r" in coords`  or `"t" in coords` prevents unnecessary warnings from being raised. The reason this works is because the logical negation of those checks, `not ("r" in coords or "t" in coords)` is equivalent to `coords == "z"` which is precisely the condition for when we replace `footgun_grid` with a better grid.

Reply:
Without these extra checks you'd get warnings when computing anything when the input grid is not a high resolution quadrautre grid

Reply:
And it still works for things like
```python
@register_compute_fun(
    name="chi",
    label="\\chi",
    units="Wb",
    units_long="Webers",
    description="Poloidal flux (normalized by 2pi)",
    dim=1,
    params=[],
    transforms={"grid": []},
    profiles=[],
    coordinates="r",
    data=["chi_r", "rho"],
    resolution_requirement="r",
)
def _chi(params, transforms, profiles, data, **kwargs):
    chi_r = transforms["grid"].compress(data["chi_r"])
    chi = cumtrapz(chi_r, transforms["grid"].compress(data["rho"]), initial=0)
    data["chi"] = transforms["grid"].expand(chi)
    return data
```

Reply:
Having written this out, I now think it's probably better to change:
`"r" in req and "r" in coords` to `"r" in req and (coords not in {"z", ""})`

Reply:
So right now, `coordinates` is used to mean two related but different things:
1. How the function varies as you move in coordinate space
2. What coordinates need to be well resolved to evaluate it

I like the idea of `resolution_requirements` replacing the second function of `coordinates` so that we only ever have to check `resolution_requirements` when computing stuff, and `coordinates` can just be for metadata etc.

Reply:
Ok I have updated all compute functions with resolution requirements

Reply:
we can't just only check `resolution_requirements`, because we also need to know how we can interpolate back to the original grid. Tha latter information resides in `coordinates` metadata